### ISSUE TITLE
Generic lock configuration

### ISSUE THREAD

**Issue by pmcatominey** [gruntwork-io/terragrunt#30]
Hi,

I'm in the process of adding support for locking via Azure Storage Blobs and have a suggestion to make it simpler to implement a new lock provider. By using a more generic config format similar to Terraform's remote state configuration with a backend name and config map.

Example:
```hcl
dynamoDbLock = {
  stateFileId = "my-app"
}

# Would become

lock = {
  backend = "dynamodb"
  config {
    state_file_id = "my-app"
  }
}
```

This would allow lock-providing packages to expose a factory function like [in Terraform](https://github.com/hashicorp/terraform/blob/master/state/remote/azure.go#L15), which would assume the role of `FillDefaults` and `Validate`.

I've also noticed that keys in the `.terragrunt` file use camelCase rather than `snake_case` like Terraform, is there a reason for this?

I'll continue to work on my Azure lock in the current code style based off the DynamoDB implementation.

**Comment by brikis98** [comment]
@pmcatominey Azure support would be awesome!

I think your approach with a more generic config map is a good one. It not only makes it easier to support more lock providers, but it would also make the Terragrunt config, both for locking and state management more consistent. I'm totally in support of this change, even though it is backwards incompatible.

I think the reason for camel case is that a) that's the default in Go and b) with `hcl.Decode`, we could deserialize directly into the Go-idiomatic variables without any changes. That said, if the HCL library supports Go tags to specify custom field names with snake case, then I'm definitely in favor of changing to snake case for the `.terragrunt` so that it is idiomiatic HCL.

**Comment by pmcatominey** [comment]
Generic locks were merged in #31, casing in the HCL config will be addressed by #32, after which this issue can be closed.
